[Tooling] Fix some initial issues with screenshot generation

AliSoftware opened the pull request:
See paaHJt-3sE-p2#state-of-wp-jp-ios for all the details

 - The `JetpackScreenshotGeneration.swift` file was still containing some `extension BaseScreen { … }`, but the `BaseScreen` type got deleted a while ago. I [had to delete](https://github.com/wordpress-mobile/WordPress-iOS/commit/c3f9f80b2d66083ba6b688e7cefae02b64421104) that `extension` to make the target compile.
 - The list of devices to run the raw screenshots on — defined in the screenshots lane in `fastlane/lanes/screenshots.rb` — were outdated: the simulator names used were no longer part of the default simulators shipped with the latest Xcode we use. So I [had to pick new ones](https://github.com/wordpress-mobile/WordPress-iOS/commit/e84a4440160e5c2976ccbfb4e36cb0f483120e46).
 - The `SnapshotHelper.swift` file provided by `fastlane` was outdated. I [had to run fastlane snapshot update to make fastlane update it](https://github.com/wordpress-mobile/WordPress-iOS/commit/759baa2f3ccb3975827a656fdfb53d3952b106b3) to a newer version compatible with the `fastlane` version we’re using nowadays.
 - I’ve also [added a `skip_clean:true` option to the screenshots lane to skip the deletion of DerivedData](https://github.com/wordpress-mobile/WordPress-iOS/commit/61ae3a9c8559e4b66144e4e7fc8da8a37a3c7d74) at the start of the lane, so that we can avoid having to do a clean build which takes forever every time we do an iterative tweaking while fixing the screenshots. That makes those little trial-and-error iterations much faster!

---

Even after fixing all the above issues and finally being able to compile and make the screenshots lane run, the UI Tests are all failing (Error code 65), because the actual screens of the app have changed a lot since last time we updated the UI Test used for screenshots — so it couldn’t find the proper `XCUIElements` it expected on screen anymore.

So this PR is not enough to make the Screenshot UI Tests run, and those will need to be updated and fixed separately. But the changes in this PR are still a way forward.

AliSoftware committed:
`fastlane download_jetpack_promo_strings`

AliSoftware committed:
Fix `BaseScreen` type not existing anymore

AliSoftware committed:
`fastlane snapshot update`

To update the helper file and make it match the current version of fastlane we use

AliSoftware committed:
Add `screenshots` lane option to avoid clean build

While it makes sense to deletes DerivedData when generating the final screenshots, to be sure to avoid relying on potentially stale cache, this means that creating screenshots will always need to rebuild the whole app from scratch first.

When iterating by trial-and-error — typically when someone needs to tweak the screenshots or debug the lanes until they get to the result they want — it's a big waste of time to do a clean build on each iteration while we typically only changed a quick thing to tweak the setup and retry to see how it looks.

Also, when we will ultimately run this lane on CI every night or so to ensure the Screenshots UI Test targets still compile and work (just to test the lane, as opposed to actually re-generate all the screenshots for all the locales with the intention of updating them in App Store), re-using the DerivedData from a build of the app made from a previous CI step would also be useful.

mokagio commented on .bundle/config:
```diff
@@ -1,4 +1,4 @@
 BUNDLE_PATH: "vendor/bundle"
 BUNDLE_JOBS: "16"
-BUNDLE_WITHOUT: "screenshots"
+BUNDLE_WITH: "screenshots"
```
I see this comes from a dedicated commit, a573f06a0c150dd4c174fd537ec3031ff7e1c50b.

What benefits do you see in forcing the `screenshots` group to be installed by everyone?

The group require RMagick, which in turn requires ImageMagick. Is it necessary for _all_ developers to have to install and maintain ImageMagick?

mokagio replied:
FWIW, this currently fails in CI (see [this test branch](https://buildkite.com/automattic/wordpress-ios/builds/8178#0181375d-9ce4-4885-ba02-4454ff7ef804/400-547)):

![image](https://user-images.githubusercontent.com/1218433/172098050-6ef52b8f-33ad-439f-b9b7-45d5a9d78dbf.png)

But, that's something we'll have to fix regardless of whether `.bundle/config` contains `BUNDLE_WITH` or `BUNDLE_WITHOUT: "screenshots"` if we want to run the screenshots in Buildkite.

AliSoftware replied:
This was not intentional at all and will need to be reverted. Nicely spotted 👍

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -16,11 +16,20 @@
   #
   desc 'Generate localised screenshots'
   lane :screenshots do |options|
     sh('bundle exec pod install')
```
Possibly out of scope, but it caught my eye because I've been [removing `sh` calls recently](https://github.com/woocommerce/woocommerce-android/pull/6680): This could become a call to the `cocoapods` action. 🤷‍♂️

AliSoftware replied:
Done in 153e5b875fc58253049d1061effb137406a5bfd0

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -16,11 +16,20 @@
   lane :screenshots do |options|
     sh('bundle exec pod install')
-    FileUtils.rm_rf(DERIVED_DATA_PATH)
+    FileUtils.rm_rf(DERIVED_DATA_PATH) unless options[:skip_clean]
```
Did you consider using a positive verb parameter, i.e. `clean`?

```suggestion
    FileUtils.rm_rf(DERIVED_DATA_PATH) if options[:clean]
```

FWIW, that's what `gym` and `scan` do. From their respective docs:

```
| clean                  | Should the project be  | GYM_CLEAN              | false                |
|                        | cleaned before         |                        |                      |
|                        | building it?           |                        |                      |
```

```
| clean                  | Should the project be  | SCAN_CLEAN             | false                  |
|                        | cleaned before         |                        |                        |
|                        | building it?           |                        |                        |
```

I guess one reason might be to signify that "clean" is the default behavior. In that case, `skip_clean` makes it clear that one is opt-in out of the default behavior.

Personally, I find positive parameters make the logic easier to follow (do this if "should do this" true; vs. do this unless "should not do this" true).

I'm just asking out of curiosity, IMHO either option is find in this context.

AliSoftware replied:
> I guess one reason might be to signify that "clean" is the default behavior.

That's exactly why I chose "skip" here. When a developer will be asked to work on updating the screenshots and they will work on it, they might not know that this option exists and would then call the lane without any specific option (i.e. rely on the default) to generate the final screenshots as well… and we want those to be a clean build

That narrative could change though once we manage to make the final screenshot generation be done and run by Buildkite. In that case, we could prefer the default — when developers will iterate a lot on the screenshots until they're happy with final design — to be incremental build, and enforce clean just when on CI when the final images will be expected to be generated from… but since we're not completely there yet… I'm not sure 🤷

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -16,11 +16,20 @@
+  #
+  #   `bundle exec fastlane screenshots language:fr-FR skip_clean:true`
+  #
+  # That way, it uses incremental builds instead of clean builds (faster iterations), and only generates screenshots for one language
```
Faster iterations 🎉 💪

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -61,10 +70,10 @@
-          'iPad Pro (12.9-inch) (3rd generation)'
+          'iPhone 13 Pro Max',
+          'iPhone 11 Pro',
+          'iPad Pro (12.9-inch) (4th generation)',
```
[CI couldn't find this Simulator](https://buildkite.com/automattic/wordpress-ios/builds/8180#01813776-56f5-4f18-8894-0fc249570125/890-8212):

![image](https://user-images.githubusercontent.com/1218433/172103370-d60ab22a-3de8-41e4-aa88-fd0ce19c651a.png)

I'm not sure whether we should create it or change simulator type.

AliSoftware replied:
Yeah I thought about creating it as well (`xcrun simctl create` and so on, actually Fastlane has a FourFlusher gem — and probably some actions as well — for that so there's probably something that can be done as well there… I'm surprised that action taking the screenshots doesn't create the sims by default already btw…).

Will definitively need to explore that.

AliSoftware committed:
Replace `sh('pod install')` -> `cocoapods` action

AliSoftware committed:
Create simulators if they don't exist

AliSoftware commented:
@mokagio

 - I've rebased the whole PR on top of `trunk`, to keep it in sync and not too far behind, and also to do some cleanup of the commits to address your feedback (i.e. remove the one which did `bundle install --with screenshots`, …)
 - I've implemented the simulator creation (only if they don't exist yet, to avoid creating too many duplicate simulators with multiple runs especially on frequent iterations) in 80fd9a077e480badb1cd8a2aebf15dc1239ff84e
 - I've also took the liberty to rebase your test branch `fix-screenshots-test-simulators` on top of my rebase. [Let's see how the Buildkite build goes for it now… 🤞 ](https://buildkite.com/automattic/wordpress-ios/builds/8417)

AliSoftware commented on fastlane/lanes/screenshots.rb:
```diff
@@ -2,6 +2,19 @@
+#
+SCREENSHOT_SIMULATORS = [
+  'iPhone Xs Max', # 6.5in - 2688x1242 @ 458 ppi
+  'iPhone 8 Plus', # 5.5in - 1920x1080 @ 401 ppi -- !!! FIXME: `canvas_size` in `fastlane/screenshots.json` does not match ([1242, 2208])
```
I've decided not to address this in this PR, because:
 - Fixing that would require updating a lot of coordinates (offsets and sizes) in `screenshots.json`, and potentially to also update the device frame PNG used for that device… and to do that properly we need to have the Screenshots UI Tests (producing the raw screenshots) working first, to then test how the promo screenshots using that `screenshots.json` would need to be adjusting… so this will definitively be for a later PR
 - Also, the aspect ratio of 2208/1242 = 1920/1080 = 16/9 is the same for both models… so maybe that does not matter that much after all (i.e. the raw screenshots, generated from an iPhone 8 Plus with dimensions 1920x1080, will _not_ be distorted and will keep a coherent aspect ratio when inserted into the frame made for iPhone 7 Plus in the promo screenshot)

Still, it would be nice to keep that consistent next time we make the screenshots UI Tests work again (hopefully soon, see pbArwn-4Pq-p2#comment-6053) and then potentially update the list of devices and the `screenshots.json` files with more modern devices (but we need to have the UI Test and raw screenshots work first as they are a necessary input for iterating on the promo screenshots)

AliSoftware commented:
[Dammit](https://buildkite.com/automattic/wordpress-ios/builds/8417#0181733f-6ccb-434f-9738-ffa049d410f3/7272-7378):
```


[20:37:09]: ▸ xcodebuild: error: Unable to find a destination matching the provided destination specifier:
--
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, OS:15.5, name:iPhone 8 Plus }
  | [20:37:09]: ▸ 	The requested device could not be found because multiple devices matched the request.
  | [20:37:09]: ▸ 	Available destinations for the "WordPressScreenshotGeneration" scheme:
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:F2A642D4-67B1-44C1-B2F5-3D00B29F1824, OS:15.5, name:iPad (9th generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:9FF18CF7-2FE0-4243-82E8-6ECF3A769975, OS:15.5, name:iPad Air (5th generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:AEA6FCAC-3118-4FD5-AA2B-8E797182375E, OS:15.5, name:iPad Pro (9.7-inch) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:8B221416-052E-4062-A472-4D95D729C857, OS:15.5, name:iPad Pro (11-inch) (3rd generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:187AFF9F-74B8-4700-9992-A803F1BD2CA3, OS:15.5, name:iPad Pro (12.9-inch) (2nd generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:1D901D16-CD8B-4D5D-BC8B-8B0CF4014140, OS:15.5, name:iPad Pro (12.9-inch) (3rd generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:ADC6BBE9-049E-4903-AC48-DE9C9700DE72, OS:15.5, name:iPad Pro (12.9-inch) (5th generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:454C806B-05A7-426A-9A23-34F44115FB0F, OS:15.5, name:iPad mini (6th generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:415BC35F-8BC9-46BA-B640-0DC90192DECF, OS:15.5, name:iPhone 8 }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:2C0526F1-6036-44EA-9C8D-117A35EED912, OS:15.5, name:iPhone 8 Plus }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:9985097F-5C46-40A8-BCE8-C4E8F59FB96F, OS:15.5, name:iPhone 8 Plus }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:8FDE8627-6074-4A34-A876-BB326C0CF430, OS:15.5, name:iPhone 11 }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:CD8C13AD-46C9-4AA1-9200-E09CAC252BE9, OS:15.5, name:iPhone 11 Pro }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:41CB717C-767E-476F-9716-7E07C6B806FD, OS:15.5, name:iPhone 11 Pro Max }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:6A4D4988-6BB0-41CD-8366-57F68FD176EC, OS:15.5, name:iPhone 12 }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:6A18A974-8721-4859-A9F7-A78B9869400D, OS:15.5, name:iPhone 12 Pro }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:42907D8D-2018-4B52-AA6B-BAD53464C0F8, OS:15.5, name:iPhone 12 Pro Max }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:2E5743D5-8AB4-4511-9E84-D4BFE2D9D039, OS:15.5, name:iPhone 12 mini }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:2EE5C900-3BF0-415C-9995-A630BFBAF39A, OS:15.5, name:iPhone 13 }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:1460E854-CC14-40E0-96F9-9CB490E3A105, OS:15.5, name:iPhone 13 Pro }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:0B86964C-CA40-47D2-832C-7FA4B9FB5CDD, OS:15.5, name:iPhone 13 Pro Max }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:A69A7B37-5D7C-4788-B44C-38C3943EF408, OS:15.5, name:iPhone 13 mini }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:CCD662A8-06DB-496C-98DB-7C8B4DA1E09D, OS:15.5, name:iPhone SE (3rd generation) }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:EA2BB17E-D3D3-45BA-B996-F3EE50F27245, OS:15.5, name:iPhone Xs Max }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:44B8E4FD-2448-4F7C-8A5D-07E3C0CD7890, OS:15.5, name:iPod touch (7th generation) }
  | [20:37:09]: ▸ 	Ineligible destinations for the "WordPressScreenshotGeneration" scheme:
  | [20:37:09]: ▸ 		{ platform:iOS, id:dvtdevice-DVTiPhonePlaceholder-iphoneos:placeholder, name:Any iOS Device }
  | [20:37:09]: ▸ 		{ platform:iOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-iphonesimulator:placeholder, name:Any iOS Simulator Device }
```
And indeed earlier in the logs we can see that during `create_missing_simulators_for_screenshots` it created all for simulators, without detecting that an `iPhone 8 Plus` already existed…

I'm not sure why it created that `iPhone 8 Plus` for `iOS 15.5` while there was already an existing one. When I tested the `create_missing_simulators_for_screenshots` locally on my Mac it correctly detected simulators that were already matching that name and skipped creating them in those cases… so not sure why it didn't do the same on CI as expected 🤔

AliSoftware committed:
Create missing sims only after Xcode invocation

Because when the CI VM starts, Xcode has never been launched yet, and the list of simulators returned by `xcrun simctl` is thus basically empty. But a default set of simulators is created by Xcode after the first launch of `xcodebuild` and invocation of `scan`.

So we need to create the missing simulators only **after** that first invocation of `xcodebuild`/`scan`, not before, to avoid creating simulators that would be also created by Xcode as part of the default set and end up with multiple simulators with the same device name, which would result in `capture_ios_screenshots` to fail to launch the UI test with a "The requested device could not be found because multiple devices matched the request." error otherwise.

AliSoftware commented:
Ok, I've now fixed the above issue in 3040f94ec244f27cb9b9426b528eb576d2a08e1f — Successful CI run [here](https://buildkite.com/automattic/wordpress-ios/builds/8437#018181ae-af9d-4e83-b16f-ed9f2329b0a1)

TL;DR: I called the `create_missing_simulators_for_screenshots` too early, especially _before_ `xcodebuild` had ever been invoked once after the Ci VM got created. At such an early stage, because Xcode had never been launched on that VM, the list of simulators was empty, hence why my script created the 4 ones. Turns out we need to invoke that lane only _after_ at least one invocation of `xcodebuild`, so that Xcode would already have created the default set of simulators on its first invocation, and our command would only create the missing ones from that default set afterwards.

@mokagio this is now ready for another review 🙂

AliSoftware committed:
Improve logging on CI

mokagio commented on WordPress/JetpackScreenshotGeneration/JetpackScreenshotGeneration.swift:
```diff
@@ -85,18 +85,6 @@ class JetpackScreenshotGeneration: XCTestCase {
     }
 }
 
-extension BaseScreen {
-    @discardableResult
-    func thenTakeScreenshot(_ index: Int, named title: String) -> Self {
-        let mode = XCUIDevice.inDarkMode ? "dark" : "light"
-        let filename = "\(index)-\(mode)-\(title)"
-
-        snapshot(filename)
-
-        return self
-    }
-}
-
```
Thank you for this cleanup.

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -192,4 +211,33 @@
+        Actions.execute_action(step_name) { UI.message "A simulator for device type #{device_type_id} already exists" }
+      else
+        res = sh('xcrun', 'simctl', 'create', device_name, device_type_id, step_name: step_name)
+        # res.split("\n").find { |line| line.match?(/^[0-9A-F-]+$/) }&.chomp # UUID of the created simulator
```
I don't usually endorse checking in commented code, but this is definitely useful to keep here for future debugging use 😄

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -192,4 +211,33 @@
+
+  # Create simulators needed for the `screenshots` lane if they don't exist yet
+  #
+  # Note: Be sure to call this lane after `xcodebuild` has been invoked at least once in the VM / local machine
+  # because on a fresh VM where Xcode has never launched yet, there might be no simulator at all, but as soon as
+  # Xcode is launched for the first time it will create a set of defaut simulators.
+  #
+  # We only want to create the missing simulators for screenshots if they don't exist in the default set created
+  # after the first `xcodebuild` launch, otherwise if we create a simulator while it will also be created by Xcode
+  # later, we'll end up with `The requested device could not be found because multiple devices matched the request.`
+  # when we try to run `capture_ios_screenshots`
```
It would be nice to find away to automate a guard against this occurrence, but I think it's safe to leave here because it relates to specialized tooling that 1) few people work with and 2) is automated, therefore easier to run in a way that accounts for this note 👍

AliSoftware replied:
Another idea I thought about was to create the simulators in all cases, but then (1) either reference them with their UUID (hence line 239 😛) when passing it to the `devices:`, or alternatively (2) create them with distinctive names like `"StoreScreenshots - iPhone 8 Plus"` (to avoid risk of "multiple matching devices…" error) and passing those to `devices:`. And optionally delete them in some `ensure` block at the end of the lane.

But I wasn't sure if that would risk impacting the name of the generated files (raw PNG screenshots) created by fastlane during each run of the UITests, so preferred to keep it at that for the first iteration.

(I might try experimenting with it again later depending on how the rest of the project goes… 🤞 )

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -16,11 +29,21 @@
   # @option [String] scheme (default: "WordPressScreenshotGeneration") The name of the scheme to use to run the screenshots
   # @option [String] output_directory (default: `${PWD}/screenhots`) The directory to generate the screenshots to
   # @option [Array<String>] language (default: the list of Mag16 locales) The subset of locales to generate the screenshots for
+  # @option [Boolean] skip_clean Skip the deletion of DerivedData before starting. Useful to speed up iterating while adjusting screenshots until we're happy with the final results, for example
+  #
+  # @note When you're working on updating the screenshots with new design, and might need to run this lane multiple time to do some
+  # trial and errors to adjust the results until you're happy with it, you might want to run this lane with the following options:
+  #
+  #   `bundle exec fastlane screenshots language:fr-FR skip_clean:true`
+  #
+  # That way, it uses incremental builds instead of clean builds (faster iterations), and only generates screenshots for one language
+  # (which is usually enough to check that the design and screens being captured look correct while iterating).
   #
   desc 'Generate localised screenshots'
   lane :screenshots do |options|
-    sh('bundle exec pod install')
-    FileUtils.rm_rf(DERIVED_DATA_PATH)
+    cocoapods # pod install
```
+1 for using the built-in action

mokagio reviewed (approved):
Happy to see a green build in CI. I think that's all we needed from this PR: to make sure we have a stable workspace on which to execute the next iterations.
